### ISSUE TITLE
Consider making Router independent of aspnet.

### ISSUE THREAD

**Issue by dazinator** [aspnet/Home#2830]
# TLDR

Would you consider opening up `Routing` to be used generically (i.e also in Non web apps), by removing the dependency on `Microsoft.AspNetCore.*` and promoting the actual Routing (i.e the bit that evaluates routes to find a matching handler) into an "Microsoft.Extensions.Routing" ?

### More details behind motivation..
The reason I ask is just that a generic `Router` implementation could be a useful standardised component, with uses outside of just asp.net core website - for example, in an `Xamarin.Forms` app it could be used to implement a navigation service that can route to an appropriate handler which could be provided by a miriad of MVVM frameworks out there.

```
var fooArg = new Foo();

// uses Routing behind the scenes to route to a handler for this route. 
await NavigationService.NavigateAsync("/about/{parameter}", fooArg ); 

[MvvmCrossRoute("about")]
public class AboutViewModel
{ 
  // mvvm passes nav params to Prepare();
   public AboutViewModel() 
   {

   }

   public override void Prepare(Foo parameter)
   {
        // receive and store the parameter here
        _myObject = parameter;
    }

 // other mvvmcross related lifecyle methods below..
}

[PrismRoute("newstuff")]
public class NewStuffModel
{ 

  // prism passes nav params to `OnNavigatedTo`
   public MyViewModel() 
   {

   }
  public void OnNavigatedTo(NavigationParameters parameters)
  {
    Foo foo  = parameters["parameter"];       
  }
 // other prism related lifecyle methods below..

}
```

In terms of usage in an xamarin forms app targeting netstadard, it would be fairly similar to asp.net core:

```

// Adds an INavigationService to the container, configured with a "pipeline" analagous to
// asp.net core RequestPipeline, except a NavigationRequest flows through it. 
// The `middleware` in the pipeline is in `IRouter` implementation,  and there is a Default Handler 
// provided by an Mvvm framework  of choice, and a `NavigationRequest` flows through that pipeline. 
// Routes can be configured via a RouteBuilder or the Mvvm framework of choice might also discover 
// them using Route attributes (like mvc's mega attribute route)

services.AddNavigation((routingOptions)=>{
   routingOptions.UseMvvmCross((routes)=>{
     // routes.AddRoute() etc.
   });  
});

```
Obviously this is just an idea of one potential example usage!

**Comment by davidfowl** [comment]
@dazinator it would be better if you forked the routing project and used the pieces you think are valuable. This team would likely be the wrong team to design a generic routing system for all application types. That should be reverse engineered from concrete usage proven and implementations.

**Comment by dazinator** [comment]
@davidfowl - ok, will close this
